plugins/example-code-snippets/src/examples/resolvers/source-parsers/regex-snippet-source-parser.ts

```diff
@@ -0,0 +1,92 @@
+import {SnippetSourceParser} from './snippet-source-parser';
+import {ExampleSnippetId, ExampleSnippetType} from '../../examples';
+import * as fs from 'fs';
+
+export interface RegexSnippetTypeOptions {
+  sourceFiles: Array<string>;
+  startRegex: (snippetId: ExampleSnippetId) => RegExp;
+  endRegex: (snippetId: ExampleSnippetId) => RegExp;
+  numLeadingSpacesToStrip: number;
+}
+
+export interface RegexSnippetSourceParserOptions {
+  snippetTypeParseOptions: Map<ExampleSnippetType, RegexSnippetTypeOptions>;
+}
+
+export class RegexSnippetSourceParser implements SnippetSourceParser {
+  private readonly snippetTypeParseOptions: Map<
+    ExampleSnippetType,
+    RegexSnippetTypeOptions
+  >;
+  private readonly snippetTypeContent: Map<ExampleSnippetType, Array<string>>;
+
+  constructor(options: RegexSnippetSourceParserOptions) {
+    this.snippetTypeParseOptions = options.snippetTypeParseOptions;
+    this.snippetTypeContent = new Map();
+    for (const [
+      snippetType,
+      snippetOptions,
+    ] of this.snippetTypeParseOptions.entries()) {
+      this.snippetTypeContent.set(
+        snippetType,
+        snippetOptions.sourceFiles
+          .map(f => fs.readFileSync(f).toString())
+          .join('\n')
+          .split('\n')
+      );
+      console.log(snippetOptions);
+    }
+  }
+  parseSourceForSnippet(
+    snippetType: ExampleSnippetType,
+    snippetId: ExampleSnippetId
+  ): string | undefined {
+    const options = this.snippetTypeParseOptions.get(snippetType);
+    if (options === undefined) {
+      return undefined;
+    }
+
+    const startRegex = options.startRegex(snippetId);
+    const endRegex = options.endRegex(snippetId);
+
+    const content = this.snippetTypeContent.get(snippetType);
+    if (content === undefined) {
+      return undefined;
+    }
+
+    const contentLinesIterator = content[Symbol.iterator]();
+
+    let nextLine = contentLinesIterator.next();
+    while (!nextLine.done) {
+      const line = nextLine.value;
+      if (line.match(startRegex)) {
+        return this.captureUntilEndOfSnippet(
+          contentLinesIterator,
+          endRegex,
+          options.numLeadingSpacesToStrip
+        );
+      }
+      nextLine = contentLinesIterator.next();
+    }
+
+    return undefined;
+  }
+
+  private captureUntilEndOfSnippet(
+    remainingLines: IterableIterator<string>,
+    endRegex: RegExp,
+    numLeadingSpacesToStrip: number
```

Comment:
We could infer the amount of leading whitespace to strip from the snippets themselves, ie $$min_{line \in snippet}\\{\text{amount of leading whitespace in line}\\}$$

Otherwise we should rename this to `numLeadingWhitespaceCharsToStrip` because it sounds like space symbols as opposed to any whitespace (tabs).

Reply:
yeah this parameter is definitely going to change in future PRs as we add more repos in.  It will probably become a higher-order function argument so that each SDK can specify its own rules for normalizing the example lines.